policy/policy.go

```diff
@@ -214,94 +214,134 @@ func (p Policy) Verify(ctx context.Context, opts ...VerifyOption) (map[string][]
-		if verifiedStatement.Statement.PredicateType != attestation.CollectionType {
-			log.Debugf("(policy) skipping statement: predicate type is not a collection (%v)", verifiedStatement.Statement.PredicateType)
-			continue
+func (step Step) checkFunctionaries(statements []source.CollectionVerificationResult, trustBundles map[string]TrustBundle) []source.CollectionVerificationResult {
+	for i, statement := range statements {
+		// Check that the statement contains a predicate type that we accept
+		if statement.Statement.PredicateType != attestation.CollectionType {
+			statements[i].Errors = append(statement.Errors, fmt.Errorf("predicate type %v is not a collection predicate type", statement.Statement.PredicateType))
 		}
 
-		for _, verifier := range verifiedStatement.Verifiers {
-			for _, functionary := range step.Functionaries {
-				if err := functionary.Validate(verifier, trustBundles); err != nil {
-					log.Debugf("(policy) skipping verifier: %w", err)
-					continue
-				} else {
-					collections = append(collections, verifiedStatement)
+		if len(statement.Verifiers) > 0 {
+			for _, verifier := range statement.Verifiers {
+				for _, functionary := range step.Functionaries {
+					if err := functionary.Validate(verifier, trustBundles); err != nil {
+						statements[i].Warnings = append(statement.Warnings, fmt.Sprintf("failed to validate functionary of KeyID %s in step %s: %s", functionary.PublicKeyID, step.Name, err.Error()))
+						continue
+					} else {
+						statements[i].ValidFunctionaries = append(statement.ValidFunctionaries, verifier)
+					}
 				}
 			}
+		} else {
+			statements[i].Errors = append(statement.Errors, fmt.Errorf("no verifiers present to validate against collection verifiers"))
 		}
 	}
 
-	return collections
+	return statements
 }
 
 // verifyArtifacts will check the artifacts (materials+products) of the step referred to by `ArtifactsFrom` against the
 // materials of the original step.  This ensures file integrity between each step.
-func (p Policy) verifyArtifacts(collectionsByStep map[string][]source.VerifiedCollection) (map[string][]source.VerifiedCollection, error) {
-	acceptedByStep := make(map[string][]source.VerifiedCollection)
+func (p Policy) verifyArtifacts(resultsByStep map[string]StepResult) (map[string]StepResult, error) {
 	for _, step := range p.Steps {
-		accepted := make([]source.VerifiedCollection, 0)
-		for _, collection := range collectionsByStep[step.Name] {
-			if err := verifyCollectionArtifacts(step, collection, collectionsByStep); err == nil {
-				accepted = append(accepted, collection)
+		accepted := false
+		if len(resultsByStep[step.Name].Passed) == 0 {
+			if result, ok := resultsByStep[step.Name]; ok {
+				result.Rejected = append(result.Rejected, RejectedCollection{Reason: fmt.Errorf("failed to verify artifacts for step %s: no passed collections present", step.Name)})
+				resultsByStep[step.Name] = result
 			}
+			return resultsByStep, nil
```

Comment:
Yeah my apologies on this one, very unclear. I have now changed things so it continues if there are no passing results.

Reply:
actually gone an backtracked a bit. We want to handle an error in the case of `!ok` because that implies that there was a problem fetching a step result (which should definitely be there)